File: ConfusingNumbersII.cs
// Time Complexity :O(2^n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : NA, premium question
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// 1) We can calculate all valid numbers from hashmap 
// 2) Check if it is valid confusing number and if its less than n
// 3) add to result


Dictionary<int, int> dict;
int count;
public int confusingNumbers(int n)
{
    if (n == 0)
        return 0;

    dict.Add(0, 0);
    dict.Add(1, 1);
    dict.Add(6, 9);
    dict.Add(8, 8);
    dict.Add(9, 6);

    dfsConfusingNumber(n, 0);

    return count;

}

private void dfsConfusingNumber(int n, long currNumber)
{
    //base
    //if curr number is greater than n, we dont those need numbers
    if (currNumber > n)
        return;

    //logic
    //check if curr number is valid confusing number
    if (isValidConfusingNumbers(currNumber))
        count++;

    //get new number to check to call dfs
    foreach (var key in dict.Keys)
    {
        long newNumber = currNumber * 10 + key;
        if(newNumber != 0)
            dfsConfusingNumber(n, newNumber);
    }
}

private bool isValidConfusingNumbers(long currNumber)
{
    long original = currNumber;
    long reverseNumber = 0;
    //we reverse the original number
    while(currNumber > 0)
    {
        reverseNumber = reverseNumber * 10 + dict[(int)currNumber % 10)];
        currNumber = currNumber / 10;
    }

    return currNumber != original;
}

File: ConfusingNumbersII.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

// Time Complexity : O(5^k) k is the number of digits in the number n;
// Space Complexity : O(5+k)
// Did this code successfully run on Leetcode : Yes

public class ConfusingNumbersII {
    HashMap<Integer, Integer> map ;
    List<Integer> confusingNumbers;
    int result;
    public int confusingNumberII(int n) {
        confusingNumbers = new ArrayList<>();
        map = new HashMap<>();
        map.put(0,0);
        map.put(1,1);
        map.put(6,9);
        map.put(8,8);
        map.put(9,6);

        helper(0l,n);
        System.out.println(confusingNumbers);
        return result;
    }

    private void helper(long currNum, int n){
        if(currNum > n) return;
        if(isConfusing(currNum)){
            confusingNumbers.add((int)currNum);
            result++;
        }
        for(int i: map.keySet()){
            long nextNum = currNum*10 + i;
            if(nextNum != 0)
                helper(nextNum,n);
        }
    }

    private boolean isConfusing(long num){
        long temp = num;
        long currNum=0;
        while(num > 0){
            int lastDigit = (int)num%10;
            num = num/10;
            currNum = currNum*10 + map.get(lastDigit);
        }
        return currNum != temp;
    }
}